cartographer_rviz/cartographer_rviz/submaps_display.cc

```diff
@@ -107,49 +107,48 @@ void SubmapsDisplay::processMessage(
       continue;
     }
-    const auto& trajectory = trajectories_[trajectory_id].second;
-    const auto it = trajectory.find(submap_entry.submap_index);
-    if (it != trajectory.end() &&
+
+    const auto& submaps = trajectories_[trajectory_id]->submaps();
+    const auto it = submaps.find(submap_entry.submap_index);
+    if (it != submaps.end() &&
         it->second->version() > submap_entry.submap_version) {
       // Versions should only increase unless Cartographer restarted.
       trajectories_.clear();
       break;
     }
   }
+
   using ::cartographer::mapping::SubmapId;
   std::set<SubmapId> listed_submaps;
   for (const ::cartographer_ros_msgs::SubmapEntry& submap_entry : msg->submap) {
     const SubmapId id{submap_entry.trajectory_id, submap_entry.submap_index};
     listed_submaps.insert(id);
     while (id.trajectory_id >= static_cast<int>(trajectories_.size())) {
-      trajectories_.push_back(Trajectory(
-          ::cartographer::common::make_unique<::rviz::Property>(
-              QString("Trajectory %1").arg(id.trajectory_id), QVariant(),
-              QString("List of all submaps in Trajectory %1.")
-                  .arg(id.trajectory_id),
-              submaps_category_),
-          std::map<int, std::unique_ptr<DrawableSubmap>>()));
-    }
-    auto& trajectory_category = trajectories_[id.trajectory_id].first;
-    auto& trajectory = trajectories_[id.trajectory_id].second;
-    if (trajectory.count(id.submap_index) == 0) {
-      trajectory.emplace(
-          id.submap_index,
-          ::cartographer::common::make_unique<DrawableSubmap>(
-              id, context_->getSceneManager(), trajectory_category.get(),
+      trajectories_.push_back(
+          ::cartographer::common::make_unique<DrawableTrajectory>(
+              id.trajectory_id, submaps_category_, context_->getSceneManager(),
               visibility_all_enabled_->getBool()));
     }
-    trajectory.at(id.submap_index)
-        ->Update(msg->header, submap_entry, context_->getFrameManager());
+
+    auto& trajectory = trajectories_[id.trajectory_id];
+    const auto& submaps = trajectory->submaps();
+
+    if (submaps.count(id.submap_index) == 0) {
+      trajectory->AddSubmap(id);
+    }
+    trajectory->UpdateSubmap(id.submap_index, msg->header, submap_entry,
+                             context_->getFrameManager());
   }
+
   // Remove all submaps not mentioned in the SubmapList.
   for (size_t trajectory_id = 0; trajectory_id < trajectories_.size();
        ++trajectory_id) {
-    auto& trajectory = trajectories_[trajectory_id].second;
-    for (auto it = trajectory.begin(); it != trajectory.end();) {
+    auto& trajectory = trajectories_[trajectory_id];
+    auto& submaps = trajectory->submaps();
+    for (auto it = submaps.begin(); it != submaps.end();) {
       if (listed_submaps.count(
               SubmapId{static_cast<int>(trajectory_id), it->first}) == 0) {
-        it = trajectory.erase(it);
+        it = submaps.erase(it);
```

Comment:
This is iffy... You are modifying data that you are not owning. This violates separation of concerns and is really hard to reason about.